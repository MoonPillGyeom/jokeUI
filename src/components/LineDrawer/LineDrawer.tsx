import React, { useEffect, useRef, useState } from 'react';
import Pallet from '../Pallet/Pallet';
import LineDrawerUtils from '../../utils/LineDrawerUtils';

export default function LineDrawer() {
  const canvasRef = useRef<HTMLCanvasElement>(null);
  const contextRef = useRef<CanvasRenderingContext2D | null>(null);
  const [penWidth, setPenWidth] = useState(1);
  const [color, setColor] = useState('black');
  const { handleStartDrawing, handleDraw, handleStopDrawing } = LineDrawerUtils({ contextRef });
  useEffect(() => {
    const canvas = canvasRef.current;
    const context = canvas?.getContext('2d');
    if (context) {
      context.lineCap = 'round';
      context.strokeStyle = `${color}`;
      context.lineWidth = Number(`${penWidth}`);
      contextRef.current = context;
    }
  }, [penWidth, color]);

  return (
    <div>
      <Pallet contextRef={contextRef} setColor={setColor} setPenWidth={setPenWidth} />
      <canvas
        style={{ backgroundColor: '#eee' }}
        ref={canvasRef}
        width={1000}
        height={800}
        onMouseDown={handleStartDrawing}
        onMouseMove={handleDraw}
        onMouseUp={handleStopDrawing}
      />
    </div>
  );
}
