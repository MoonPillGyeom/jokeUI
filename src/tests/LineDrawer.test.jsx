import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import LineDrawer from '../components/LineDrawer';

describe('LineDrawer Component', () => {
  test('canvas to see line', () => {
    const { getByTestId } = render(<LineDrawer />);
    const canvas = getByTestId('canvas');

    fireEvent.mouseDown(canvas, { clientX: 50, clientY: 50 });
    fireEvent.mouseMove(canvas, { clientX: 100, clientY: 100 });
    fireEvent.mouseUp(canvas);

    const context = canvas.getContext('2d');
    expect(context.__getDrawCalls()).toMatchSnapshot(); // jest-canvas-mock을 통해 호출 내용 확인
  });
  test('Clike AllEareaButton to clear canvas', () => {
    const { getByTestId, getByText } = render(<LineDrawer />);

    const canvas = getByTestId('canvas');

    // 캔버스에 선을 그리는 시뮬레이션
    fireEvent.mouseDown(canvas, { clientX: 50, clientY: 50 });
    fireEvent.mouseMove(canvas, { clientX: 100, clientY: 100 });
    fireEvent.mouseUp(canvas);

    const eraserButton = getByText('Allearea');
    fireEvent.click(eraserButton);

    const context = canvas.getContext('2d');

    expect(context.__getDrawCalls()).toMatchSnapshot(); // __getDrawCalls()함수를 이용해 clearRect가 실행되었는지 확인
  });
});
