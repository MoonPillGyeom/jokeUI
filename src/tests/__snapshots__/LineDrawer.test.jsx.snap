// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineDrawer Component Clike AllEareaButton to clear canvas 1`] = `
[
  {
    "props": {
      "path": [
        {
          "props": {},
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  {
    "props": {
      "path": [
        {
          "props": {},
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  {
    "props": {
      "height": 800,
      "width": 1000,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
]
`;

exports[`LineDrawer Component canvas to see line 1`] = `
[
  {
    "props": {
      "path": [
        {
          "props": {},
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  {
    "props": {
      "path": [
        {
          "props": {},
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
]
`;
